README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# DrugSeekR

<!-- badges: start -->
<!-- badges: end -->

## Description

`DrugSeekR` is an R package that analyzes genomic data in the format of
MAF files to identify potential drug targets for personalized medicine
through mutational frequency analysis, gene annotation, and drug target
matching. It aims to provide a user-friendly and comprehensive workflow
for finding personalized treatments based on genomic samples from
patients. It leverages existing tools and databases, such as maftools,
biomaRt, and Drug Gene Interaction Database, to perform various steps of
the analysis. The package was developed using
`R version 4.3.1 (2023-06-16)`, Platform: `x86_64-apple-darwin20`
running under `macOS: Monterey 12.6.8`.

## Installation

You can install the development version of DrugSeekR from
[GitHub](https://github.com/) with:

``` r
require("devtools")
devtools::install github("this-wei/DrugSeekR", build vignettes = TRUE)
library("DrugSeekR")
```

To run the shinyApp: Under construction

## Overview

`DrugSeekR` is designed to analyze MAF file inputs. MAF files for
analysis can be found at the repository of the The Cancer Genome Atlas
Program (TCGA), which provides publicly available genomic, epigenomic,
transcriptomic, and proteomic data.

The link is provided here:
(<https://portal.gdc.cancer.gov/repository?filters=%7B%22op%22%3A%22and%22%2C%22content%22%3A%5B%7B%22op%22%3A%22in%22%2C%22content%22%3A%7B%22field%22%3A%22files.access%22%2C%22value%22%3A%5B%22open%22%5D%7D%7D%2C%7B%22op%22%3A%22in%22%2C%22content%22%3A%7B%22field%22%3A%22files.data_format%22%2C%22value%22%3A%5B%22maf%22%5D%7D%7D%5D%7D>)

Note you must unzip and provide the path to the maf.gz file to the
maftools::read.maf function.

``` r
ls("package:DrugSeekR")
browseVignettes("DrugSeekR")
```

`DrugSeekR` contains 4 functions: 1. ***calculateMutationFrequency***
for calculating the frequency and percentage of observed mutations
within a sample. 2. ***annotateGenes*** for annotating the genes with
information from the ensembl database and returns the top genes with the
highest number of mutations. Note this function must use dbplyr version
2.3.4 due to a known bug 3. ***getGeneDrugTargets*** for finding the
most druggable genes and the drugs that target them from the Drug Gene
Interaction Database. 4. ***plotDruggableGenes*** for ploting the number
of drugs for each gene using ggplot2.

![](./inst/extdata/bch_workflow.png)

## Contributions

The author of the package is Wei Jin Luo. The author wrote the
calculate_mutation_frequency function, which calculates the frequency
and percentage of observed mutations within a sample. The
calculate_mutation_frequency function makes use of the maftools R
package to get a summary of mutation data.

The author wrote the annotate_genes function, which annotates the genes
with information from the ensembl database and returns the top genes
with the highest number of mutations. The functions make use of the
biomaRt R package to query the ensembl database.

The author wrote the get_gene_drug_targets function, which finds the
most druggable genes and the drugs that target them from the Drug Gene
Interaction Database. The functions makes use of the maftools R package
to query the Drug Gene Interaction Database.

The author wrote the plot_druggable_genes function, which plots the
number of drugs for each gene. The plot_druggable_genes function makes
use of the ggplot2 R package to create the plot.

## References

- Mayakonda A, Lin DC, Assenov Y, Plass C, Koeffler HP (2018). Maftools:
  efficient and comprehensive analysis of somatic variants in cancer.
  Genome Research. <http://dx.doi.org/10.1101/gr.239244.118>

- Steffen Durinck, Paul T. Spellman, Ewan Birney and Wolfgang Huber
  (2009). Mapping identifiers for the integration of genomic datasets
  with the R/Bioconductor package biomaRt. Nature Protocols 4, 1184-1191
  .

- Steffen Durinck, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor,
  Alvis Brazma and Wolfgang Huber (2005). BioMart and Bioconductor: a
  powerful link between biological databases and microarray data
  analysis. Bioinformatics 21, 3439-3440

- Griffith, M., Griffith, O., Coffman, A. et al (2013). DGIdb: mining
  the druggable genome. Nat Methods 10, 1209–1210.
  <https://doi.org/10.1038/nmeth.2689>

- H. Wickham (2016). ggplot2: Elegant Graphics for Data Analysis.
  Springer-Verlag New York

## Acknowledgements

This package was developed as part of an assessment for 2022 BCB410H:
Applied Bioinformatics course at the University of Toronto, Toronto,
CANADA. DrugSeekR welcomes issues, enhancement requests, and other
contributions. To submit an issue, use the GitHub issues.


NEWS.md
# DrugSeekR 0.1.0

* Initial submission


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(DrugSeekR)

test_check("DrugSeekR")


tests/testthat/test-annotateGenes.R
library(DrugSeekR)

geneData <- data.frame(
  Hugo_Symbol = c("BRCA1", "BRCA2", "TP53", "KRAS", "EGFR"),
  Frame_Shift_Del = c(10, 15, 20, 25, 30),
  Frame_Shift_Ins = c(5, 10, 15, 20, 25),
  In_Frame_Del = c(15, 20, 25, 30, 35),
  Missense_Mutation = c(20, 25, 30, 35, 40),
  total = c(50, 70, 90, 110, 130)
)

returnData <- annotateGenes(geneData)
returnDataNumber <- annotateGenes(geneData, number = 2)

test_that("annotateGenes works as expected", {
  # Test the output type
  expect_type(returnData, "list")

  # Test the output length
  expect_length(returnData, 2)

  # Test the output names
  expect_named(returnData, c("geneAnnotations", "numberOfMutations"))

  # Test the output values
  expect_setequal(returnData$geneAnnotations$hgnc_symbol, c("BRCA1", "BRCA2", "TP53", "KRAS", "EGFR"))

  expect_equal(returnDataNumber$geneAnnotations$hgnc_symbol, c("EGFR", "KRAS"))
})


tests/testthat/test-getGeneDrugTargets.R
library(DrugSeekR)

geneData <- data.frame(
  Hugo_Symbol = c("BRCA1", "BRCA2", "TP53", "KRAS", "EGFR"),
  Frame_Shift_Del = c(10, 15, 20, 25, 30),
  Frame_Shift_Ins = c(5, 10, 15, 20, 25),
  In_Frame_Del = c(15, 20, 25, 30, 35),
  Missense_Mutation = c(20, 25, 30, 35, 40),
  total = c(50, 70, 90, 110, 130)
)

test_that("getGeneDrugTargets returns the expected output", {
  result_no_number <- getGeneDrugTargets(geneData)
  result_number <- getGeneDrugTargets(geneData, number = 2)

  expect_type(result_no_number, "list")
  expect_length(result_no_number, 2)
  expect_named(result_no_number,
               c("mostDruggableGenes", "drugsThatTargetTheseGenes"))

  expect_equal(names(result_no_number$mostDruggableGenes),
               c("EGFR", "KRAS", "TP53", "BRCA1", "BRCA2"))
  expect_equal(names(result_number$mostDruggableGenes),
                  c("EGFR", "KRAS"))

})






R/get_drugs.R
#' Get gene drug targets
#'
#' This function takes a data frame of gene data and returns a list of
#' the most druggable genes and information on the drugs that target them.
#' Gene druggability is found from the Drug Gene Interaction Database
#'  referenced below
#'
#' @param geneData A data frame of gene data that must column Hugo_Symbol.
#' @param number An integer specifying the number of most druggable genes to return.
#' Default is NULL which gets all genes.
#'
#' @return A list with two elements:
#' \itemize{
#'   \item mostDruggableGenes: A named vector of the most druggable genes and
#'    their frequency in the drug database.
#'   \item drugsThatTargetTheseGenes: A data frame of the drugs that target
#'    the most druggable genes with columns Gene, interaction_claim_source,
#'   interaction_types, drug_claim_name, drug_claim_primary_name, drug_name,
#'    drug_chembl_id, PMIDs
#' }
#'
#' @examples
#' # Example 1:
#' # Use gene data provided from the maftools package
#' laml.maf <- system.file('extdata', 'tcga_laml.maf.gz', package = 'maftools')
#' laml <- maftools::read.maf(maf = laml.maf)
#' geneSummary <- maftools::getGeneSummary(laml)
#'
#' getGeneDrugTargets(geneSummary)
#'
#' @references
#' Mayakonda A, Lin DC, Assenov Y, Plass C, Koeffler HP. 2018.
#'  Maftools: efficient and comprehensive analysis of somatic variants in cancer.
#'   Genome
#' Research. http://dx.doi.org/10.1101/gr.239244.118
#'
#' Griffith, M., Griffith, O., Coffman, A. et al.
#' DGIdb: mining the druggable genome. Nat Methods 10, 1209–1210 (2013).
#'  https://doi.org/10.1038/nmeth.2689
#'
#' @export
#' @import maftools
getGeneDrugTargets <- function(geneData, number = NULL) {
  drugs <- maftools::drugInteractions(maf = NULL,
                                      genes = geneData$Hugo_Symbol,
                                      drugs = TRUE)
  sortedGenes <- sort(table (drugs$Gene), decreasing = TRUE)

  if (!is.null(number)) {
    mostDruggableGenes <- sortedGenes[1:number]
    topGenes <- names(mostDruggableGenes)[1:number]
    topDrugs <- drugs[drugs$Gene %in% topGenes, ]
  }
  else {
    mostDruggableGenes <- sortedGenes
    topDrugs <- drugs
  }

  return(list(mostDruggableGenes = mostDruggableGenes,
              drugsThatTargetTheseGenes = topDrugs))
}



R/mutationStatistics.R
#' Calculates Mutation Frequency and Percentage
#'
#' A function that calculates
#' the frequency of observed mutations within a sample and their percentages
#'
#' @param geneData A list of genetic data returned from the maftools
#'    getGeneSummary function
#'
#'
#' @return returns a list of two elements: the number of mutations for
#'    each gene and the mutation percentage for each gene.
#'
#' @examples
#' # Example 1:
#' # Using maftools maf data available with package
#' laml.maf <- system.file('extdata', 'tcga_laml.maf.gz', package = 'maftools')
#' laml <- maftools::read.maf(maf = laml.maf)
#' geneSummary <- maftools::getGeneSummary(laml)
#'
#' # Calculate Mutation Frequency and Percentage
#' mutationStatistics <- calculateMutationFrequencies(geneSummary)
#' mutationStatistics
#'
#' @references
#' Mayakonda A, Lin DC, Assenov Y, Plass C, Koeffler HP. 2018. Maftools:
#' efficient and comprehensive analysis of somatic variants in cancer. Genome
#' Research. http://dx.doi.org/10.1101/gr.239244.118
#'
#' @export
#' @import maftools
calculateMutationFrequencies <- function(geneData) {
  mutationSums <- colSums(geneData[, -1])

  total <- sum(geneData$total)
  percentages <- mutationSums / total * 100
  percentages <- round(percentages, 2)

  return(list(numberOfMutations = mutationSums,
              mutationByPercentage = percentages))
}


R/visualize_drug_targets.R
#' Plot druggable genes
#'
#' This function takes a named vector of drug frequencies for each gene and
#'  returns a bar plot of the number of drugs for each gene.
#'
#' @param drugFrequencies A named vector of drug frequencies for each gene.
#'
#' @return A ggplot object of a bar plot of the number of drugs for each gene.
#'
#' @examples
#' # Example 1:
#' # Use gene data provided from the maftools package
#' laml.maf <- system.file('extdata', 'tcga_laml.maf.gz', package = 'maftools')
#' laml <- maftools::read.maf(maf = laml.maf)
#' geneSummary <- maftools::getGeneSummary(laml)
#'
#' drugTargets <- DrugSeekR::getGeneDrugTargets(geneSummary)
#' plotDruggableGenes(drugTargets$mostDruggableGenes)
#'
#' @references
#' H. Wickham. ggplot2: Elegant Graphics for Data Analysis.
#'  Springer-Verlag New York, 2016.
#'
#' @export
#' @import ggplot2
plotDruggableGenes <- function(drugFrequencies) {
  gene <- NULL
  drugCount <- NULL
  geneCounts <- as.data.frame(drugFrequencies)

  colnames(geneCounts) <- c("gene", "drugCount")
  ggplot2::ggplot(geneCounts, aes(x = gene, y = drugCount)) +
    geom_bar(stat = "identity", fill = "steelblue") +
    labs(title = "Number of drugs for each gene",
         x = "Gene",
         y = "Drug count") +
    theme_minimal()
}


R/annotate_genes.R
#' Annotate genes
#'
#' This function takes a data frame of gene data and returns a list of gene
#' annotations and number of mutations for the top genes.
#' NOTE: This function requires a stable internet connection to run as it
#' queries the ensembl database for gene annotations.
#'
#' @param geneData A data frame or list of gene data and must contain column
#' names Hugo_Symbol and total.
#' @param number An integer specifying the number of top genes to annotate.
#' Default is NULL which will annotate all genes.
#' @param refDataset A string representing the species that you want to
#' identify the genes.
#' Available datasets can be listed using biomaRt::listDatasets()
#' Default is "hsapiens_gene_ensembl".
#'
#' @return A list with two elements:
#' \itemize{
#'   \item geneAnnotations: A data frame of gene annotations from biomaRt
#'   with columns hgnc_symbol, ensembl_gene_id, chromosome_name, start_position,
#'   end_position, and description.
#'   \item numberOfMutations: A data frame of the top genes and
#'   their number of mutations.
#' }
#'
#' @examples
#' # Example 1
#' # Use gene data provided from the maftools package
#' laml.maf <- system.file('extdata', 'tcga_laml.maf.gz', package = 'maftools')
#' laml <- maftools::read.maf(maf = laml.maf)
#' geneSummary <- maftools::getGeneSummary(laml)
#' annotateGenes(geneData = geneSummary)
#'
#' @references
#' Mapping identifiers for the integration of genomic datasets with the
#' R/Bioconductor package biomaRt.
#' Steffen Durinck, Paul T. Spellman, Ewan Birney and
#' Wolfgang Huber, Nature Protocols 4, 1184-1191 (2009).
#'
#' BioMart and Bioconductor: a powerful link between biological databases and
#' microarray data analysis. Steffen Durinck, Yves Moreau, Arek Kasprzyk, Sean
#' Davis, Bart De Moor, Alvis Brazma and Wolfgang Huber,
#' Bioinformatics 21, 3439-3440 (2005).
#'
#' @export
#' @import biomaRt
annotateGenes <- function(geneData,
                          number = NULL,
                          refDataset = "hsapiens_gene_ensembl") {

  sorted <- geneData[order(-geneData$total), ]
  # Check if user specified a number
  if (!is.null(number)) {
    top <- sorted[1:number, ]
    genes <- top$Hugo_Symbol
  }
  else {
    genes <- sorted$Hugo_Symbol
  }

  ensembl <- biomaRt::useEnsembl(biomart = "genes", dataset = refDataset)

  filters <- "hgnc_symbol"
  attributes <- c("hgnc_symbol",
                  "ensembl_gene_id",
                  "chromosome_name",
                  "start_position",
                  "end_position",
                  "description")
  results <- biomaRt::getBM(filters = filters,
                            attributes = attributes,
                            values = genes,
                            mart = ensembl)

  return(list(geneAnnotations = results, numberOfMutations = sorted))
}
